### PR TITLE
RDDFC-408: Extend Fission Router to Support Sticky Cookies

### PR DESCRIPTION
**Pull request by OS-tariqabbasi**

### CODE REVIEW

File: charts/fission-all/values.yaml [hunk 1]
```diff
@@ -192,6 +192,9 @@ router:
   ## unTapService is called to free up the resources once the function invocation is done.
   ##
   unTapServiceTimeout: 3600s
+  ## isValidTimeout is the timeout used in the request context of isValid.
+  ## isValid is called to check if a function service referenced by a stickiness cookie is still valid.
+  isValidTimeout: 30s
```
Comment (OS-saraghuravi) [inline charts/fission-all/values.yaml:197]: Can we move this inside `stickiness` instead of root? we only use this when stickiness is enabled right?

**Review by OS-saraghuravi — APPROVED** [review]
LGTM

File: charts/fission-all/values.yaml (hunk 1 above)
Comment (OS-tariqabbasi) [reply charts/fission-all/values.yaml:197]: I wanted to group it with the other timeouts. It is currently only used when stickiness is enabled however, someone might want to use it for something else later (and the timeout is generic).

File: pkg/router/functionHandler.go [hunk 2]
```diff
@@ -440,8 +478,70 @@ func (fh *functionHandler) tapService(fn *fv1.Function, serviceURL *url.URL) {
 	fh.executor.TapService(fn.ObjectMeta, fn.Spec.InvokeStrategy.ExecutionStrategy.ExecutorType, *serviceURL)
 }
 
+func (fh functionHandler) parseStickinessCookie(request *http.Request) (*stickinessCookie, *http.Cookie, error) {
+	var stickinessCookie stickinessCookie
+	cookie, err := request.Cookie(STICKINESS_COOKIE)
+	if err != nil {
+		fh.logger.Debug("failed to get stickiness cookie value", zap.Error(err))
```
Comment (OS-vascogomes) [inline pkg/router/functionHandler.go:485]: Are this "errors" not important enough to be Errors? Should we keep them with Debug?

File: pkg/router/functionHandler.go [hunk 3]
```diff
@@ -440,8 +478,70 @@ func (fh *functionHandler) tapService(fn *fv1.Function, serviceURL *url.URL) {
...
+				fh.logger.Debug("expiring invalid stickiness cookie")
+				_, err = fh.invalidateStrictStickinessCookie(responseWriter, cookie)
+				if err != nil {
+					fh.logger.Error("error writing invalid stickiness cookie response", zap.Error(err))
+				}
+				return
+			}
+
+		} else {
+			fh.logger.Debug("failed to parse stickiness cookie, will be ignored", zap.Error(err))
```
Comment (OS-vascogomes) [inline pkg/router/functionHandler.go:542]: I think we should log some error, if not here, inside the parseStickinessCookie. Do you agree?

File: pkg/router/functionHandler.go (hunk 2 above)
Comment (OS-tariqabbasi) [reply pkg/router/functionHandler.go:485]: This one will happen on every new client / connection so would keep as debug

Comment (OS-tariqabbasi) [reply pkg/router/functionHandler.go:485]: If it fails to base64 decode / json unmarshal then that should potentially be an error however I don't think we can log the cookie contents in this case (potential security risk)

File: pkg/router/functionHandler.go (hunk 3 above)
Comment (OS-tariqabbasi) [reply pkg/router/functionHandler.go:542]: This one will happen a lot as well, a couple of the debug calls in the function are good candiates for error level logging

File: pkg/router/functionHandler.go (hunk 2 above)
Comment (OS-tariqabbasi) [reply pkg/router/functionHandler.go:485]: I have increased the relevant log levels to WARN and removed the cookie contents from the log, let me know if you agree

**Review by OS-vascogomes — APPROVED** [review]

**Pull request merged**